src/SonarScanner.Shim/PathHelper.cs

```diff
@@ -22,40 +22,86 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using SonarQube.Common;
 
 namespace SonarScanner.Shim
 {
     public static class PathHelper
     {
-        public static bool IsPartOfAProject(string filePath, IEnumerable<string> projectPaths)
+        public static string WithTrailingDirectorySeparator(this DirectoryInfo directory)
         {
-            return projectPaths.Any(projectPath => IsInFolder(filePath, projectPath));
+            if (directory == null)
+            {
+                throw new ArgumentNullException(nameof(directory));
+            }
+
+            if (directory.FullName.EndsWith(Path.DirectorySeparatorChar.ToString()) ||
+                directory.FullName.EndsWith(Path.AltDirectorySeparatorChar.ToString()))
+            {
+                return directory.FullName;
+            }
+
+            if (directory.FullName.Contains(Path.AltDirectorySeparatorChar))
+            {
+                return directory.FullName + Path.AltDirectorySeparatorChar;
+            }
+
+            return directory.FullName + Path.DirectorySeparatorChar;
         }
 
-        public static bool IsInFolder(string filePath, string folder)
+        public static bool IsInDirectory(FileInfo file, DirectoryInfo directory)
         {
-            var normalizedPath = Path.GetDirectoryName(Path.GetFullPath(filePath));
-            return normalizedPath.StartsWith(folder, StringComparison.OrdinalIgnoreCase);
+            var normalizedDirectoryPath = directory.WithTrailingDirectorySeparator();
+            return file.FullName.StartsWith(normalizedDirectoryPath, FileInfoEqualityComparer.ComparisonType);
         }
 
-        public static string GetCommonRoot(IEnumerable<string> paths)
+        public static DirectoryInfo GetCommonRoot(IEnumerable<DirectoryInfo> paths)
         {
-            var projectDirectoryParts = paths
-                .Select(p => p.Split(Path.DirectorySeparatorChar))
-                .ToList();
-
-            if (projectDirectoryParts.Count == 0)
+            if (paths == null)
             {
-                return string.Empty;
+                return null;
             }
 
+            var projectDirectoryParts = paths
+                .Select(GetParts)
+                .ToList();
+
             var commonParts = projectDirectoryParts
-                .OrderBy(p => p.Length)
+                .OrderBy(p => p.Count)
                 .First()
                 .TakeWhile((element, index) => projectDirectoryParts.All(p => p[index] == element))
                 .ToArray();
 
-            return string.Join(Path.DirectorySeparatorChar.ToString(), commonParts);
+            switch (commonParts.Length)
+            {
+                case 0:
+                    return null;
+                case 1:
+                    return new DirectoryInfo(commonParts[0]);
```

Comment:
This case isn't strictly necessary - _Path.Combine_ works with an array of a single element.

Reply:
I had a weird case when testing but maybe it was linked to something else I fixed. I will make another test without this special handling.